### PR TITLE
Feature/v2 updates

### PR DESCRIPTION
**Pull request by fivetran-jamie**
**Are you a current Fivetran customer?** 

fivetran made PR

**What change(s) does this PR introduce?** 

- revamps campaign and line item reports
- converts ad adapter model into url report
- adds account_report, promoted_tweet_report, and keyword_report
- incorporates passthrough columns into report tables
- adds schema tests to all end models 
- updates README
- documents new tables and columns
- adds to changelog and decisionlog 
- TODO AFTER APPROVING: regen docs
- 
**Did you update the CHANGELOG?** 

- [x] Yes

**Does this PR introduce a breaking change?**

- [x] Yes (please provide breaking change details below.)
- [ ] No  (please provide an explanation as to how the change is non-breaking below.)

**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 

- [x] Yes

**Is this PR in response to a previously created Bug or Feature Request**

- [ ] Yes, Issue/Feature [link bug/feature number here]
- [ ] No 

**How did you test the PR changes?** 

 

- [x] CircleCi  
- [x] Local (please provide additional testing details below)

**Select which warehouse(s) were used to test the PR**
 

- [x] BigQuery
- [x] Redshift
- [x] Snowflake
- [x] Postgres
- [x] Databricks
- [ ] Other (provide details below)

**Provide an emoji that best describes your current mood**
 
🎦 

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

### CODE REVIEW

File: models/twitter.yml [hunk 1]
```diff
@@ -1,96 +1,398 @@
...
-    description: Each record represents the daily performance of a line item at the URL level.
+  - name: twitter_ads__campaign_report
+    description: >
+      Each record in this table represents the daily performance of ads at the account and campaign level, within a placement in Twitter.
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - date_day
+            - campaign_id 
+            - placement
```
Comment (fivetran-joemarkiewicz) [inline models/twitter.yml:12]: The table description states that the grain exists as campaign -> account. However, we are only testing the uniqueness of the three columns you included above. If the hierarchy stands to be consistent across the board, can we also test against the uniqueness of the grain we are stating.

Comment (fivetran-joemarkiewicz) [reply models/twitter.yml:12]: Same comment above for the other end models in this package.

File: models/twitter_ads__account_report.sql [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    from {{ var('promoted_tweet_report') }}
+),
+
+rollup_report as (
+
+    select 
+        date_day,
+        account_id,
+        placement,
+        sum(clicks) as clicks,
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/twitter_ads__account_report.sql:20]: Same comments for other models in this package.

File: models/twitter_ads__account_report.sql [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+rollup_report as (
+
+    select 
+        date_day,
+        account_id,
+        placement,
+        sum(clicks) as clicks, 
+        sum(impressions) as impressions,
+        sum(spend) as spend,
+        sum(spend_micro) as spend_micro,
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/twitter_ads__account_report.sql:23]: Same comments for other models in this package.

File: packages.yml [hunk 4]
```diff
@@ -1,3 +1,7 @@
 packages:
-- package: fivetran/twitter_ads_source
-  version: [">=0.4.0", "<0.5.0"]
+# - package: fivetran/twitter_ads_source
+#   version: [">=0.4.0", "<0.5.0"]
+  # - local: ../dbt_twitter_source
+  - git: https://github.com/fivetran/dbt_twitter_source.git
+    revision: feature/v2-updates
+    warn-unpinned: false
```
Comment (fivetran-joemarkiewicz) [inline packages.yml:7]: Friendly reminder to switch before merge

File: README.md [hunk 5]
```diff
@@ -1,90 +1,142 @@
...
+
+| **Model**                | **Description**                                                                                                                                |
+| ------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------- |
+| [twitter_ads__account_report](https://fivetran.github.io/dbt_twitter/#!/model/model.twitter.twitter_ads__account_report)             | Each record in this table represents the daily performance at the account and ad-placement level. |
+| [twitter_ads__campaign_report](https://fivetran.github.io/dbt_twitter/#!/model/model.twitter.twitter_ads__campaign_report)            | Each record in this table represents the daily performance of ads at the account, campaign, and ad-placement level. |
+| [twitter_ads__line_item_report](https://fivetran.github.io/dbt_twitter/#!/model/model.twitter.twitter_ads__line_item_report)            | Each record in this table represents the daily performance of ads at the campaign, account, line item, and ad-placement level.|
```
Comment (fivetran-joemarkiewicz) [inline README.md:28]: Can we maybe add a similar note you have in the `twitter.yml` file for this table description that provides the relation that line item is essentially the same as ad group.

File: models/twitter.yml [hunk 6]
```diff
@@ -1,96 +1,398 @@
...
+      - name: updated_timestamp
+        description: "{{ doc('updated_at') }}"
+      - name: bid_amount
+        description: The bid amount to be associated with this line item.
+      - name: total_budget_amount
+        description: The total budget amount to be allocated to the campaign.
+      - name: target_cpa
+        description: The target cost per acquisition for the line item.
+      - name: placement
+        description: Where on Twitter the ad is being displayed. Possible values include 'ALL_ON_TWITTER', 'PUBLISHER_NETWORK', 'TWITTER_PROFILE', 'TWITTER_SEARCH', 'TWITTER_TIMELINE', and 'TAP_*', which are more granular options for `PUBLISHER_NETWORK`.
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter.yml:155]: ah yep adding to all reports

File: models/twitter.yml [hunk 7]
```diff
@@ -1,96 +1,398 @@
...
+      - name: account_name
+        description: The name of the related account.
+      - name: created_timestamp
+        description: "{{ doc('created_at') }}"
+      - name: currency
+        description: The currency all metrics for the account are set to.
+      - name: is_deleted
+        description: "{{ doc('deleted') }}"
+      - name: duration_in_days
+        description: The time period within which the frequency_cap is achieved.
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter.yml:49]: meant to remove it

File: models/twitter.yml [hunk 8]
```diff
@@ -1,96 +1,398 @@
...
+        description: The impressions for the promoted tweet + URL on that day.
       - name: spend
-        description: The amount of spend on that date.
+        description: The spend for the promoted tweet + URL on that day.
+      - name: spend_micro
+        description: The spend, in micros, for the tweet + URL on that day.
+      - name: url_clicks
+        description: The URL clicks for the promoted tweet + URL on that day.
+      - name: approval_status
+        description: The approval status of the promoted tweet.
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter.yml:311]: will include in model selecttion

File: models/twitter_ads__account_report.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter_ads__account_report.sql:23]: we'll want to use `spend` in the rollup -- i believe the other platforms don't report in micros

File: models/twitter.yml [hunk 9]
```diff
@@ -1,96 +1,398 @@
...
+        description: The name of the line item.
       - name: spend
         description: The amount of spend on that date.
+      - name: spend_micro
+        description: The amount of spend, in micros, on that date.
       - name: url_clicks
         description: The number of URL clicks on that date.
+      - name: placement
+        tests:
+          - not_null
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter.yml:107]: oh huh weird.... well just added a description to the yml for good measure ❓

File: models/twitter.yml (hunk 1 above)
Comment (fivetran-jamie) [reply models/twitter.yml:12]: added other columns to this test for each model

File: models/twitter.yml [hunk 10]
```diff
@@ -1,96 +1,398 @@
...
-  - name: twitter__campaign_report
-    description: Each record represents the daily ad performance of each campaign.
-    columns:
-      - name: daily_campaign_id
-        description: Unique ID for each record.
+      - name: placement
+        description: Where on Twitter the ad is being displayed. Possible values include 'ALL_ON_TWITTER', 'PUBLISHER_NETWORK', 'TWITTER_PROFILE', 'TWITTER_SEARCH', 'TWITTER_TIMELINE', and 'TAP_*', which are more granular options for `PUBLISHER_NETWORK`.
         tests:
-          - unique
           - not_null
```
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter.yml:37]: yep `placement` should always be populated

File: models/twitter_ads__account_report.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (fivetran-jamie) [reply models/twitter_ads__account_report.sql:20]: yep -- `Includes clicks on the URL (shortened or regular links), profile pic, screen name, username, detail, hashtags, and likes.` <- i'll add to the docs!

File: README.md (hunk 5 above)
Comment (fivetran-jamie) [reply README.md:28]: added

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-jamie thanks for addressing my comments and updating your PR. My only comment would be to add the keyword disable variable to the required section of the README. Once you make that update, these changes look good to go once all our other packages are ready to be merged.

**Pull request merged**